### PR TITLE
[WIP] wallet: use BlockFilterIndex in ScanForWalletTransactions

### PR DESCRIPTION
**Pull request by pstratem**
goal is to improve rescan performance significantly, lots more work to do

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Have you seen #15845 ? It might be easier to take that code and then fix up the two (or three) bugs.

**Comment by MarcoFalke** [comment]
Concept ACK either way

**Comment by pstratem** [comment]
@MarcoFalke I hadn't, it seems you took basically the same approach that I intend to.  I'll certainly review your PR and read the comments.

**Comment by pstratem** [comment]
this actually works for p2pkh transactions as it is now, benchmarks incoming

**Comment by pstratem** [comment]
some rescanblockchain benchmarks, from my laptop:

42 minutes on master
20 minutes using the BASIC filter
4  minutes using a filter with fixed siphash parameters and avoiding recalculating the siphashes for every block

**Comment by jonatack** [comment]
Concept ACK

**Comment by gmaxwell** [comment]
4 minutes is awesome. 20 minutes is kinda sad compared to 42 (only 2x the speed in exchange for a bunch of disk space?).  42 minutes is *a lot* faster than what I was benchmarking before, so hurrah for whatever prior change made rescanning faster.

**Review by ryanofsky — COMMENTED** [review]
Concept & approach ACK. Seems to pick up nicely where #15845 left off, updated to work for descriptor and legacy wallets

**Comment by jb55** [comment]
Concept ACK *as I grumpily wait while my wallets are rescanning*

File: src/interfaces/chain.cpp [hunk 1]
```diff
@@ -186,6 +187,20 @@ class ChainImpl : public Chain
         }
         return nullopt;
     }
+    Optional<bool> filterMatchesAny(const uint256& hash, const GCSFilter::ElementSet& filter_set) override
```
Comment (fanquake) [inline src/interfaces/chain.cpp:190]: Please [use `std::optional` in new code](https://github.com/bitcoin/bitcoin/blob/master/src/optional.h#L12).

**Comment by MarcoFalke** [comment]
Marked up for grabs, along with #15845

**Comment by MarcoFalke** [comment]
Picked up in #25957

**Pull request closed without merging**